### ISSUE TITLE
Incorrect isolate test `tests/lib_2/isolate/ping_pause_test.dart` ?

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#37787]
From [tests/lib_2/isolate/ping_pause_test.dart](https://github.com/dart-lang/sdk/blob/master/tests/lib_2/isolate/ping_pause_test.dart):
```dart
isomain1(replyPort) {
  RawReceivePort port = new RawReceivePort();
  port.handler = (v) {
    replyPort.send(v);
    if (v == 0) port.close();
  };
  replyPort.send(port.sendPort);
}

void main() {
  asyncStart();
  var completer = new Completer(); // Completed by first reply from isolate.
  RawReceivePort reply = new RawReceivePort(completer.complete);
  Isolate.spawn(isomain1, reply.sendPort).then((Isolate isolate) {
    List result = [];
    completer.future.then((echoPort) {
      reply.handler = (v) {
        result.add(v);
        if (v == 0) {
          Expect.listEquals([4, 3, 2, 1, 0], result);
          reply.close();
          asyncEnd();
        }
      };
      echoPort.send(4);
      echoPort.send(3);
      Capability resume = isolate.pause();
      var pingPort = new RawReceivePort();
      pingPort.handler = (_) {
        Expect.isTrue(result.length <= 2);
        echoPort.send(0);
        isolate.resume(resume);
        pingPort.close();
      };
      isolate.ping(pingPort.sendPort, priority: Isolate.beforeNextEvent);
      echoPort.send(2);
      echoPort.send(1);
    });
  });
}
```

This test seems to be currently just timing out on most / all configurations. 

When looking closer at this test it seems that the `main` isolate is pausing `isomain1` and after that one is being paused it sends it a ping. Though the ping is sent via `Isolate.beforeNextEvent` which causes us to enqueue it into the normal message queue, not in the OOB queue. 

Though since the isolate is paused it will not process any more messages from the normal queue. So the ping will not get answered.

Yet the `main` isolate needs the ping to get answered before it will resume the other isolate.

=> We are "deadlocked" - none of the two isolates makes progress.

I'm going to skip this test for now.

**Comment by mkustermann** [comment]
@bkonyi As you are the author of this test, could you take a look at this?

Using `Isolate.immediate` instead of `Isolate.beforeNextEvent` would fix it, but I'm not sure if that is the intend of the test.

**Comment by bkonyi** [comment]
> @bkonyi As you are the author of this test, could you take a look at this?
> 
> Using `Isolate.immediate` instead of `Isolate.beforeNextEvent` would fix it, but I'm not sure if that is the intend of the test.

I don't believe I'm the author of this test (it's copyrighted 2014 and I started at Google late 2016 :-) ). I probably migrated it during the switch to 2.0. Unfortunately, I'm not sure what the intent of the test is either.

**Comment by mkustermann** [comment]
@lrhn You seem to have introduced this test in 705b67e836470a. Do you think the test is correct?

**Comment by lrhn** [comment]
I believe the test is correct and the implementation is wrong.

The control port events are not handled on the normal event loop. 

A request for a ping `immediately` must be answered when the request is received (after all previous immediate requests from the same isolate has been handled), whether the event loop is currently handling an event or not.

A request for a ping `beforeNextEvent` must be answered when the currently executing event, if any, is done (according to the documentation of `Isolate.ping`). This should happen whether the event loop is paused or not because control messages do not go on the event loop queue.

So, if the implementation does prepend the `beforeNextEvent` events to the normal event loop, rather than checking them on the side when control comes back to the event loop, then it should be able to recognize control events and do them even while paused.
